subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/DependencyInsightReportTask.java

```diff
@@ -270,53 +322,49 @@ private void assertValidTaskConfiguration(@Nullable Configuration configuration)
         return selectedDependencies;
     }
 
-
-    private static AttributeMatchDetails match(Attribute<?> actualAttribute, Object actualValue, AttributeContainer requestedAttributes) {
+    @SuppressWarnings("unchecked")
+    private AttributeMatchDetails match(Attribute<?> actualAttribute, @Nullable Object actualValue, AttributeContainer requestedAttributes) {
+        AttributesSchemaInternal schema = (AttributesSchemaInternal) getProject().getDependencies().getAttributesSchema();
+        // As far as I could tell, the only schema ever mixed in using withProducer is PreferJavaRuntimeVariant
+        // However, that only adds disambiguation rules, which don't apply here. So this should be sufficient:
+        AttributeMatcher matcher = schema.matcher();
         for (Attribute<?> requested : requestedAttributes.keySet()) {
```

Comment:
The reason for choosing a particular variant is lost here because we don't know how the disambiguation rules played a part.

e.g., if you look at a project dependency, you'll see ~4 compatible variants. Only one of those was selected, but from looking at just the attributes, it seems like any of them should have been picked. 

Not now, but I wonder if we could expose the attribute matcher and add a method that explains the matching process with more details when given the requested attributes _and_ all of the variant's attributes.